controllers/topology/cluster_controller_test.go

```diff
@@ -0,0 +1,361 @@
+			// use the ClusterTopologyMachineDeploymentLabel to get the specific machineDeployment to compare to.
+			if topologyMD.Name != md.GetLabels()[clusterv1.ClusterTopologyMachineDeploymentLabelName] {
+				continue
+			}
+			// Assert that the correct Finalizer has been added to the MachineDeployment.
+			for _, f := range md.Finalizers {
+				// Break as soon as we find a matching finalizer.
+				if f == clusterv1.MachineDeploymentTopologyFinalizer {
+					break
+				}
+				// False if the finalizer is not present on the MachineDeployment.
+				return false
+			}
+			if !(ensureClusterTopologyOwnedLabel(md) && ensureClusterLabelName(md, cluster.Name)) {
+				log.Infof("machine deployment does not have the correct labels.")
+				return false
+			}
+			// Check replicas for the MachineDeployment.
+			if *md.Spec.Replicas != *topologyMD.Replicas {
+				log.Infof("We failed on deployment replicas%v %v", md.Spec.Replicas, topologyMD.Replicas)
+				return false
+			}
+			if *md.Spec.Template.Spec.Version != cluster.Spec.Topology.Version {
+				log.Infof("We failed on deployment version%v", md)
+				return false
+			}
+			// Check if the InfrastructureReference exists.
+			if ok := referenceExistsWithCorrectKindAndAPIVersion(&md.Spec.Template.Spec.InfrastructureRef,
+				builder.GenericInfrastructureMachineTemplateKind,
+				builder.InfrastructureGroupVersion); !ok {
+				log.Infof("We failed on infrastructure version and kind %v", md)
+				return false
+			}
+			// Check if the InfrastructureReference has the expected labels and annotations.
+			if _, ok := getAndEnsureLabelsAndAnnotations(md.Spec.Template.Spec.InfrastructureRef, cluster.Name); !ok {
+				log.Infof("We failed on infrastructure labels and annotations. %v", md)
+				return false
+			}
+			// Check if the Bootstrap reference has the expected Kind and APIVersion.
+			if ok := referenceExistsWithCorrectKindAndAPIVersion(md.Spec.Template.Spec.Bootstrap.ConfigRef,
+				builder.GenericBootstrapConfigTemplateKind,
+				builder.BootstrapGroupVersion); !ok {
+				log.Infof("We failed on bootstrap version and kind %v", md)
+				return false
+			}
+			// Check if the Bootstrap reference has the expected labels and annotations.
+			if _, ok := getAndEnsureLabelsAndAnnotations(*md.Spec.Template.Spec.Bootstrap.ConfigRef, cluster.Name); !ok {
+				log.Infof("We failed on bootstrap labels and annotations %v", md)
+				return false
+			}
+		}
+		log.Infof("MachineDeployments passed with %v", m)
+	}
+	return true
+}
+
+// getAndEnsureLabelsAndAnnotations pulls the template referenced in the ObjectReference from the API server, checks for:
+// 1) The ClusterTopologyOwnedLabel.
+// 2) The correct ClusterLabelName.
+// 3) The annotation stating where the template was cloned from.
+// The function returns the unstructured object and a bool indicating if it passed all tests.
+func getAndEnsureLabelsAndAnnotations(template corev1.ObjectReference, clusterName string) (*unstructured.Unstructured, bool) {
```

Comment:
It seems we are using assert and ensure with the same meaning, let converge to only 1
```suggestion
func getAndAssertLabelsAndAnnotations(template corev1.ObjectReference, clusterName string) (*unstructured.Unstructured, bool) {
```

Reply:
Same for other func below